### ISSUE TITLE
Reduce AppVeyor build time

### ISSUE THREAD

**Issue by ArkadiuszMichalski** [notepad-plus-plus/notepad-plus-plus#12457]
In normal mode for one commit we have 10 jobs, each takes 8 minutes on average, which gives over 1h for each commit. It needs some optimization, especially when someone does PR for files that do not require compilation, e.g. XML (from translators). The worst case is when someone doing a PR for one XML file with xxx commits one by one and blocks building for the whole day.

I have some questions and suggestions:
1. Does AppVeyor have any button to cancel the build? What I mean is that a PR author should be able to easily cancel his build if he finds that he will be making more changes anyway.
I see that cancel button might exist (https://pythonpackaging.info/_images/appveyor-fullrun-ongoing.png), but I don't know how to turn it on and what the build cancellation permissions look like.

2. There is something like [Rolling builds](https://www.appveyor.com/docs/build-configuration/#rolling-builds). If it can be set for specific branches, e.g. just for PR, it might be worth doing it. I don't know what it looks like in details because I don't see UI settings for AppVeyor.
Looks like that such an option exists:
![image](https://user-images.githubusercontent.com/2730894/200115328-51bb5b1f-c173-4e34-819d-885fb484bd0d.png)

4. Add some mode for XML type changes where only validation will be performed, not the entire compilation. I'm testing something like this here: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/12454. 

    If someone does a PR whose title will start as `(xml) Your title` (just put `(xml)` at the beginning) we start only two jobs, and only second (Debug) check syntax for XML file. For people making minor changes to XML files that don't require any compilation, this will suffice (and take a very short time).

    Btw, I don't know how to start build with just one job XML Debug, I'm still getting XML Release and XML Debug. But it's still better than full build.

5. Add to the FAQ description how make simple PR that will not block others. Details will depend on what would be approved from the above suggestions.

**Comment by rdipardo** [comment]
I think all you need is a filter, like this:

~~~yaml
skip_commits:
  files:
    - '**/*.xml'
~~~

Reference: https://www.appveyor.com/docs/how-to/filtering-commits/#commit-files-github-and-bitbucket-only

> If someone does a PR whose title will start as `(xml) Your title`

You can filter by [commit message], too; but then contributors will have to follow the new rule . . .

[commit message]:
https://www.appveyor.com/docs/how-to/filtering-commits/#commit-message

**Comment by ArkadiuszMichalski** [comment]
>I think all you need is a filter, like this:
skip_commits:
  files:
    - '**/*.xml'

Not exactly because I still want to run the build and check the XML syntax. For skip build people can add just in-build `[skip ci]` to the title but that does not solve the problem when some tests should be performed.

Btw, in the case of XML for function list, the test is more complex, so I will skip this case for now. At the moment, I only mean things like translations or themes (or others simple things).

>You can filter by [commit message](https://www.appveyor.com/docs/how-to/filtering-commits/#commit-message), too; but then contributors will have to follow the new rule . . .

Yes, but we have a fairly narrow and rigid group of people developing the translation, so it shouldn't be a problem for them to add a simple command in the title. Especially when it will be visible in the PR list that others are doing it. Note that this is optional, and if someone doesn't add it, the full build will start (which, however, is a waste of time and resources).

If this approach were accepted we could describe somewhere these short commands and other guidelines on how not to block builds unnecessarily.

**Comment by ArkadiuszMichalski** [comment]
@donho:

As far as I can remember, this is how it was looks like:

1. For changes that don't need any build at all (eg `.md,` `.txt`,), we can add anywhere in the title `[skip ci]` or `[ci skip]` or `[skip appveyor]`. This is available out of the box and we don't have to do anything (we must remember that such annotation exists and sometimes it's worth using it). 

---

2. For changes only to XML files (like translation or theme) we just need to run the validation. I did it on this way:
Define one more environment:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/0650089a9f278fa57c939c18cec947429d19ebaf/appveyor.yml#L26-L27
It will be run only when PR has title with containing `(xml)` in any place. We can determine more precisely whether it should be `(xml)` or `[xml]`, or even a mix of characters `(XML)` or `(xMl)`.
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/0650089a9f278fa57c939c18cec947429d19ebaf/appveyor.yml#L69-L73 
If such build starts, it only contains two jobs `Release` and `Debug`. For `Release` we don't do anything, just pass and return, we check syntax only in `Debug`. I don't know how to make only one job, so it has to be that way.
    
   It's also worth adding some filtering for changed files, for example to make sure that PR with such an annotation actually changes only XML files. It's best to check this with Git:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/0650089a9f278fa57c939c18cec947429d19ebaf/appveyor.yml#L76-L80
Initially, I did a simple filtration. If someone changes files other than XML, or XML files together with others (eg CPP), we return an error. This can be adjusted later if necessary.

   If someone has changed other files and still wants to run only XML validation itself, it'is enough to create one more empty commit (`--allow-empty`) with `(xml)` in the title. and push such commit.

   That's it, simple, but probably something like that is enough.

---

3. For problem with unnecessary xxx commits one by one. Something must be done here, because someone can block the build for the whole day by adding many commits. I've seen such a situation.
   
   Like I said on the [first post](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12457#issue-1434565537) this can be done by turn on [Rolling builds](https://www.appveyor.com/docs/build-configuration/#rolling-builds) and check `Only for Pull Requests`. But this also has disadvantages because it works for all PR, we don't have any filtration. If someone really needs to queue up a few commits (for few builds), they can't do it.

   We can make own version of Rolling builds (only for PR, not master), like I did it here:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/0650089a9f278fa57c939c18cec947429d19ebaf/appveyor.yml#L38
   Here's how it works:
   
   Let say we push first commit (so fist build was queued with 10 jobs).
   Let say first job was start from this first build.
   Now we add second commit (so second build was queued with 10 jobs)
 
   Each job firstly checks if newer build for that PR was added (we check last 50 entries in history). 
   https://github.com/notepad-plus-plus/notepad-plus-plus/blob/0650089a9f278fa57c939c18cec947429d19ebaf/appveyor.yml#L40
   If newer build exist all jobs from all previous builds will be canceled.  In our case, all jobs from first build will be canceled. So the first build will not be done unnecessarily, since there is a newer one.
   https://github.com/notepad-plus-plus/notepad-plus-plus/blob/0650089a9f278fa57c939c18cec947429d19ebaf/appveyor.yml#L50

   If someone really needs to queue a few builds, just add in the title `(force)`. This is better, because by default, we don't allow many builds in one PR, unless someone explicitly indicates it.

   At this point this is not perfect yet, because every job from the canceled build still needs to be running before will be stopped. Well, it doesn't take as long as doing full compilation in these jobs!

   There is a way to terminate the entire build with currently running job. This will make breaking a given build immediate (we don't need to run all jobs only for canceling them). The code below is responsible for this:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/0650089a9f278fa57c939c18cec947429d19ebaf/appveyor.yml#L52-L62

   The problem is that we need to provide an API key to do this. It cannot be done explicitly because everyone could use it. We have something like [secure variables](https://www.appveyor.com/docs/build-configuration/#secure-variables) but that won't solve the problem either:
   >However, secure variables are not decoded during Pull Request builds which prevents someone from submitting PR with malicious build script displaying those variables. In more controlled environment through with a trusted team and private GitHub repositories there is an option on General tab of project settings to allow secure variables for PRs.

   I have no idea how idealy to get around this. We could create own PHP script, send to it what we want to stop, and then use the API key directly on PHP (no one would have access to it). Of course we also need to check if the request came from our AppVeyor account.

   Even without the above improvement, canceling each unnecessary job is much better than being able to queue any number of builds without any restrictions. Again I would recommend to focus on this problem.

Summarizing everything: @donho I can do separate PR for 2 and 3 above improvements. Maybe start with 2 first?

**Comment by donho** [comment]
> Summarizing everything: @donho I can do separate PR for 2 and 3 above improvements. Maybe start with 2 first?

Yes, sure, please go ahead.

Thank you for the very detailed and refined report. 
Ideally, if we can get committed file extension, everything will be much simpler.
If it can't be done, let's mark on the title.

**Comment by ArkadiuszMichalski** [comment]
@donho
>Ideally, if we can get committed file extension, everything will be much simpler. 

We can get such infos. Maybe not when parsing `appveyor.yml`, but when the job is running. So we can't decide how many jobs will run (always all, actually 10), but we can cancel some of them. Similar approach what I described in point 3.

So in the new approach it might look like this:
1. We do not define another environment. We are using what is now, so we always start 10 jobs.
2. From a running job, we detect which files have been changed.
3. If it's just XML files, we will cancel this job and others (9 in total). Let them report as passed.
4. We do XML validation only in one job, probably in `Win32/Debug` because that's how we do it now.

We can also catch if the changed XML files comes from the function list, if so, let it do one compilation in `Win32/Debug` because it is needed to check the parser.

So here we decide when to do XML validation only, no one has to remember to add anything in the title. Of course we still can add in the future some annotation to the title, eg. `[force xml]`, which, for example, will force the validation even if we change files other than `xml`. Or `[force build]` which, for example, will force the build even if we change only the `xml` files. It will depend on what we need in practice.
Well, the default behavior will be to do only the XML validation (unless someone explicitly indicates otherwise). This is probably a more convenient solution.

The only downside to this new approach is that checking will take a bit longer. We need to run 9 jobs just to finish them right after downloading the repo. Well, it's still better than doing an unnecessary full compilation on each of them.

**Comment by donho** [comment]
@ArkadiuszMichalski 

"a bit longer" is how much (approximately) longer? 10 seconds? 1 minutes? 3 minutes?

**Comment by ArkadiuszMichalski** [comment]
@donho 

Looking at the last test:
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45262116
30s for such unnecessary job. 30x9 = 270s = 4.5min for nothing + one job for test 40s. The whole thing will take about ~5 minutes.

With 2 jobs only 70s.
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45293737

In fact, I started doing the second variant, we'll see in practice. If it doesn't work well we will try the first one. 

But tell me is this supposed to work only for PR or also for master?

**Comment by donho** [comment]
> 30s for such unnecessary job.

It's bearable largely.

> But tell me is this supposed to work only for PR or also for master?

Let's do it on only PR for the moment. If it works fine, then master maybe.

**Comment by ArkadiuszMichalski** [comment]
@donho
Hmm I was to optimistic. The work of each job takes 30 seconds, but the job preparation itself before starting it takes a bit longer. From last log: 
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45496023
For 10 jobs it takes 13 min 47 sec
For 2 jobs would be ~3min. 

If necessary we can make a hybrid. If someone doesn't give anything in the title because he forgot or doesn't know about it, we run 10 jobs, and if he gives `(xml)` we will run 2 jobs. I will experiment more.

**Comment by ArkadiuszMichalski** [comment]
I found a way to do 1 job for the variant with `(xml)` in the title.
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45502607
Now it will take even less.

I will summarize a bit later how the current implementation works. And I will also clean the PR of unnecessary things.

**Comment by ArkadiuszMichalski** [comment]
@donho
At the moment the whole thing is best explained with a few examples.

---
1
Modified files: `english.xml` or `english.xml` + `stylers.model.xml` (only `xml` files)
Commit title: `Update english.xml`
Description: run 10 jobs, 9 will be canceled after running, only `Win32/Debug` run XML validation

---
2
Modified files: `cpp.xml` or `cpp.xml` + `english.xml` (only `xml` files, but one of them comes from the `functionList` folder)
Commit title: `Update cpp.xml`
Description: run 10 jobs, 9 will be canceled after running, only `Win32/Debug` run full compilation + validation XML and function list parsers

---
3
Modified files: `english.xml` or `english.xml` + `stylers.model.xml` (only `xml` files)
Commit title: `(xml)Update english.xml` << add anywhere on the title `(xml)` annotation
Description: run 1 job`Win32/Debug` with XML validation

---
4
Modified files: `cpp.xml` or `cpp.xml` + `english.xml` (only `xml` files, but one of them comes from the `functionList` folder)
Commit title: `(xml)Update cpp.xml` << add anywhere on the title `(xml)` annotation
Description: run 1 job`Win32/Debug` with full compilation + validation XML and function list parsers

---
5
Modified files: `Notepad_plus.cpp` or `Notepad_plus.cpp` + `english.xml` (any file is non-XML)
Commit title: `Add some function`
Description: run 10 jobs with full compilation

---
6
Modified files: `Notepad_plus.cpp` or `Notepad_plus.cpp` + `english.xml` (any file is non-XML)
Commit title: `(force xml)Add some function`
Description: run 10 jobs, 9 will be canceled after running, only `Win32/Debug` run XML validation

---
7
Modified files: `Notepad_plus.cpp` or `Notepad_plus.cpp` + `english.xml` (any file is non-XML)
Commit title: `(xml)(force xml)Add some function`
Description: run 1 job`Win32/Debug` with XML validation

**Comment by donho** [comment]
@ArkadiuszMichalski 
Looks very promising!

If I do understand well, after merging #12454, for all the PR translation (xml only), no need to add `(xml)` either `(force xml)` on the commit title, and only `Win32/Debug` will be run for XML validation ?

**Comment by ArkadiuszMichalski** [comment]
@donho 
Look once again to above list, it's 1 or 2 point, without any addition to commit's title 10 jobs must be run, but 9 will be canceled after start (no compilation) and `Win32/Debug` run XML validation (or also compilation if some of `.xml` files from function list was changed). So it will take longer, about 14 minutes. It's better than the entire compilation anyway which takes ~1.5 hours.
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45496023

Only 1 job needs `(xml)` added in the title, like 3 or 4 point in above list. If the `localization/theme` maintainers start adding it, the whole thing will take ~1minute.
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45502607

**Comment by donho** [comment]
What's the difference between `(xml)` and `(force xml)` than?

**Comment by ArkadiuszMichalski** [comment]
`(xml)` - explicitly indicates that we have changed the XML file. This ensures that only `Win32/Debug` can appear in the queue. If only `xml` files has been changed it will run `xml` validation, and if e.g. `xml` file from funciotn list or `xml + cpp` files it will perform a full build and tests. The name is short `(xml)` for easier use.

`(force xml)` - this is mainly for our internal use, because e.g. when we working on the `appveyor.yml` file, we will not be able to run XML validation from such PR (this is not `xml` file). Look at 6 or 7 point above. It 's useful to somehow force this mode regardless of the changed files.

**Comment by donho** [comment]
Isn't it better to use `[xml]` as notation?

**Comment by ArkadiuszMichalski** [comment]
@donho 
I made `(xml)` to distinguish them from the built-in ones, like `[skip ci]`. But if you prefer that all of them have `[ ]` then I can change it. There will be less typing because we don't need to press `Shift`.

I made 2 new changes:

`(xml) Some changes` << if someone changed any file other than XML here, an error will always be reported. Thanks to this, you can be sure that a standalone annotation `(xml)` will pass only when someone change only XML files.
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45513260
`Changed files on this pull request require full build.`

`(xml)(force compile) Some changes` - as opposed to the above, it will force full build `Win32/Debug`. Personally, I don't have an MSC compiler and it would be nice to be able to do a quick single compilation via AppVeyor for MSC, without doing the full 10 compilation (I don't want to block others unnecessarily). I takes only ~8min.
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45513293

From your perspective `(force xml)` and `(force compile)`, if they have been used by someone in PR, they should be used before the last commit in PR that you accept. Last commit should have only `(xml)` or nothing. These two additional flags are just for helping when we working on PR (because sometimes we can test something in shorter time).

**Comment by ArkadiuszMichalski** [comment]
>1. the markers (xml) or (force xml) must be in the beginning of the titles or they can be anywhere in the title to be taken considering?

Actually they can be anywhere. Since there can be dual annotation, this is the easiest way.

>Is the PR title will be gotten dynamically each commit?
For example, if one does a PR for the translation with (xml) in title, but he/she has found the one cpp file has to change to match the translation, the contributor then commit a cpp file - what will it happen?

The list of changed files comes from whole PR (not a single commit). In practice, when we start the job, it looks like this:
https://ci.appveyor.com/project/donho/notepad-plus-plus/builds/45513293
```
Build started
git clone -q https://github.com/notepad-plus-plus/notepad-plus-plus.git C:\projects\notepad-plus-plus
git fetch -q origin +refs/pull/12454/merge:
git checkout -qf FETCH_HEAD
```
We obtain this changed list files by:
```
$files_modified = @(git diff --name-only HEAD~1)
```
This basically includes changes from all files made so far, not just from some (single) commit. I have not found a way to get the changed files strictly from a single commit.

So in practice:

`Commit1`: change file `test.xml`
List of changed files `test.xml`
For this commit XML validation can be run.

`Commit2`: change file `test2.xml`
List of changed files `test.xml, test2.xml`
For this commit XML validation can be run.

`Commit3`: change file `some.cpp`
List of changed files `test.xml, test2.xml, some.cpp` 
For this commit standard compilation (10 jobs) must be done because other files than XML was changed.

**Comment by chcg** [comment]
Maybe we should think about using github CI builds, see https://github.com/notepad-plus-plus/notepad-plus-plus/pull/7956. There real parallel builds are also available for free and the overall buildtime therefore doesn't increase with the number of jobs we add.

**Comment by ArkadiuszMichalski** [comment]
But it doesn't change the fact that XML validation should be done on 1 job (no matter how many slots are available there). I have the transition to github CI at the back of my head https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12177, but before that happen, it's worth improving what is now. Basically this PR is ready. It's worth using it, at least for a while.

**Comment by ArkadiuszMichalski** [comment]
@donho
>What I'm trying to know is: If I change the title by adding/removing (xml) between commits, if the newest commit will change its behaviour due to the change of title?

Yes because we check title directly from the last commit. So if there is a new commit, the title of that commit will be checked.

>I prefer to use [], if possible of course. Not only for the esthetic reason, but also because () can be used on the title frequently. For example: "Clean up all the config files (xml)"

Here we can use whatever syntax you prefer, even double or triple ``[[xml]]``. But it's best to use something short so people don't have to type too much. I will change the `()` to `[]`, unless you prefer something else.
